File: src/glyph.c
/*
Copyright (c) 2013, Michael Droettboom
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

The views and conclusions contained in the software and documentation are those
of the authors and should not be interpreted as representing official policies,
either expressed or implied, of the FreeBSD Project.
*/

#include "glyph.h"
#include "doc/glyph.h"

#include "bbox.h"
#include "bitmap.h"
#include "constants.h"
#include "glyph_metrics.h"
#include "outline.h"
#include "subglyphs.h"
#include "vector.h"


#define MAKE_GLYPH_GETTER(name, convert_func, member)   \
    MAKE_GETTER(Py_Glyph, name, convert_func, member)
#define DEF_GLYPH_GETTER(name) DEF_GETTER(name, doc_Glyph_ ## name)
#define GLYPH_METHOD(name) DEF_METHOD(name, Glyph)


/****************************************************************************
 Object basics
*/

static ftpy_ConstantType Py_FT_GLYPH_BBOX_ConstantType;
static ftpy_ConstantType Py_FT_GLYPH_FORMAT_ConstantType;
static ftpy_ConstantType Py_FT_RENDER_MODE_ConstantType;


static PyTypeObject Py_Glyph_Type;


static void
Py_Glyph_dealloc(Py_Glyph* self)
{
    Py_TYPE(self)->tp_clear((PyObject *)self);
    free(self->x);
    FT_Done_Glyph(self->glyph);
    Py_TYPE(self)->tp_free((PyObject*)self);
}


PyObject *
Py_Glyph_cnew(FT_GlyphSlot glyph_slot, PyObject *owner, int load_flags)
{
    Py_Glyph *self;
    FT_GlyphSlot glyph_slot_copy;
    FT_Glyph glyph;

    if (ftpy_exc(FT_Get_Glyph(glyph_slot, &glyph))) {
        return NULL;
    }

    self = (Py_Glyph *)(&Py_Glyph_Type)->tp_alloc(&Py_Glyph_Type, 0);
    if (self == NULL) {
        return NULL;
    }

    glyph_slot_copy = PyMem_Malloc(sizeof(FT_GlyphSlotRec));
    if (glyph_slot_copy == NULL) {
        Py_DECREF(self);
        return NULL;
    }
    memcpy(glyph_slot_copy, glyph_slot, sizeof(FT_GlyphSlotRec));
    self->x = glyph_slot_copy;
    self->glyph = glyph;
    Py_INCREF(owner);
    self->base.owner = owner;
    return (PyObject *)self;
}


static PyObject *
Py_Glyph_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
{
    Py_Glyph *self;

    self = (Py_Glyph *)ftpy_Object_new(type, args, kwds);
    if (self == NULL) {
        return NULL;
    }
    self->x = NULL;
    return (PyObject *)self;
}


static int
Py_Glyph_init(Py_Glyph *self, PyObject *args, PyObject *kwds)
{
    PyErr_SetString(
        PyExc_RuntimeError,
        "Glyph objects may not be instantiated directly. "
        "Use Face.load_glyph and Face.load_char instead.");
    return -1;
}


/****************************************************************************
 Getters
*/


static PyObject *face_get(Py_Glyph *self, PyObject *closure)
{
    Py_INCREF(self->base.owner);
    return self->base.owner;
}


static PyObject *metrics_get(Py_Glyph *self, PyObject *closure)
{
    return Py_Glyph_Metrics_cnew(&self->x->metrics, (PyObject *)self, self->load_flags);
}


static PyObject *linear_hori_advance_get(Py_Glyph *self, PyObject *closure)
{
    if (self->load_flags & FT_LOAD_LINEAR_DESIGN) {
        return PyInt_FromLong(self->x->linearHoriAdvance);
    } else {
        return ftpy_PyFloat_FromFT_FIXED(self->x->linearHoriAdvance);
    }
}


static PyObject *linear_vert_advance_get(Py_Glyph *self, PyObject *closure)
{
    if (self->load_flags & FT_LOAD_LINEAR_DESIGN) {
        return PyInt_FromLong(self->x->linearVertAdvance);
    } else {
        return ftpy_PyFloat_FromFT_FIXED(self->x->linearVertAdvance);
    }
}


MAKE_GLYPH_GETTER(advance, ftpy_Py_Vector_FromVector, &self->x->advance);


static PyObject *format_get(Py_Glyph *self, PyObject *closure)
{
    return Py_Constant_cnew(&Py_FT_GLYPH_FORMAT_ConstantType, self->x->format);
}


static PyObject *bitmap_get(Py_Glyph *self, PyObject *closure)
{
    if (self->glyph->format == FT_GLYPH_FORMAT_BITMAP) {
        return Py_Bitmap_cnew(
            &((FT_BitmapGlyph)self->glyph)->bitmap);
    } else {
        PyErr_SetString(PyExc_AttributeError, "Glyph has no bitmap");
        return NULL;
    }
}


MAKE_GLYPH_GETTER(bitmap_left, PyLong_FromLong, self->x->bitmap_left);
MAKE_GLYPH_GETTER(bitmap_top, PyLong_FromLong, self->x->bitmap_top);


static PyObject *outline_get(Py_Glyph *self, PyObject *closure)
{
    if (self->glyph->format == FT_GLYPH_FORMAT_OUTLINE) {
        return Py_Outline_cnew(&((FT_OutlineGlyph)self->glyph)->outline);
    } else {
        PyErr_SetString(PyExc_AttributeError, "Glyph has no outline");
        return NULL;
    }
}


MAKE_GLYPH_GETTER(subglyphs, Py_SubGlyphs_cnew, self);
MAKE_GLYPH_GETTER(lsb_delta, PyLong_FromLong, self->x->lsb_delta);
MAKE_GLYPH_GETTER(rsb_delta, PyLong_FromLong, self->x->rsb_delta);


static PyGetSetDef Py_Glyph_getset[] = {
    DEF_GLYPH_GETTER(face),
    DEF_GLYPH_GETTER(metrics),
    DEF_GLYPH_GETTER(linear_hori_advance),
    DEF_GLYPH_GETTER(linear_vert_advance),
    DEF_GLYPH_GETTER(advance),
    DEF_GLYPH_GETTER(format),
    DEF_GLYPH_GETTER(bitmap),
    DEF_GLYPH_GETTER(bitmap_left),
    DEF_GLYPH_GETTER(bitmap_top),
    DEF_GLYPH_GETTER(outline),
    DEF_GLYPH_GETTER(subglyphs),
    DEF_GLYPH_GETTER(lsb_delta),
    DEF_GLYPH_GETTER(rsb_delta),
    {NULL}
};


/****************************************************************************
 Methods
*/


static PyObject*
Py_Glyph_get_cbox(Py_Glyph* self, PyObject* args, PyObject* kwds) {
    unsigned int bbox_mode = FT_GLYPH_BBOX_UNSCALED;
    FT_BBox bbox;
    double scale = 1.0;

    const char* keywords[] = {"bbox_mode", NULL};

    if (!PyArg_ParseTupleAndKeywords(
            args, kwds, "|I:get_cbox", (char **)keywords,
            &bbox_mode)) {
        return NULL;
    }

    FT_Glyph_Get_CBox(self->glyph, bbox_mode, &bbox);

    if (bbox_mode == FT_GLYPH_BBOX_SUBPIXELS ||
        bbox_mode == FT_GLYPH_BBOX_GRIDFIT) {
        scale = 1.0 / 64.0;
    }

    return Py_BBox_cnew(&bbox, scale);
}


static PyObject*
Py_Glyph_render(Py_Glyph* self, PyObject* args, PyObject* kwds) {
    int render_mode = FT_RENDER_MODE_NORMAL;
    double x = 0;
    double y = 0;
    FT_Vector origin;

    const char* keywords[] = {"render_mode", "origin", NULL};

    if (!PyArg_ParseTupleAndKeywords(
            args, kwds, "|i(dd):render", (char **)keywords,
            &render_mode, &x, &y)) {
        return NULL;
    }

    origin.x = TO_F26DOT6(x);
    origin.y = TO_F26DOT6(y);

    if (ftpy_exc(
            FT_Glyph_To_Bitmap(&self->glyph, render_mode, &origin, 1))) {
        return NULL;
    }

    return Py_Bitmap_cnew(
        &((FT_BitmapGlyph)self->glyph)->bitmap);
}


static PyMethodDef Py_Glyph_methods[] = {
    GLYPH_METHOD(get_cbox),
    GLYPH_METHOD(render),
    {NULL}  /* Sentinel */
};


/****************************************************************************
 Setup
*/


static PyTypeObject Py_Glyph_Type;


static PyTypeObject Py_FT_GLYPH_BBOX_Type;
#define GLYPH_BBOX_CONST(name) DEF_CONST(name, FT_GLYPH_BBOX)
static constant_def FT_GLYPH_BBOX_constants[] = {
    GLYPH_BBOX_CONST(UNSCALED),
    GLYPH_BBOX_CONST(SUBPIXELS),
    GLYPH_BBOX_CONST(GRIDFIT),
    GLYPH_BBOX_CONST(TRUNCATE),
    GLYPH_BBOX_CONST(PIXELS),
    {NULL}
};


static PyTypeObject Py_FT_GLYPH_FORMAT_Type;
#define GLYPH_FORMAT_CONST(name) DEF_CONST(name, FT_GLYPH_FORMAT)
static constant_def FT_GLYPH_FORMAT_constants[] = {
    GLYPH_FORMAT_CONST(NONE),
    GLYPH_FORMAT_CONST(COMPOSITE),
    GLYPH_FORMAT_CONST(BITMAP),
    GLYPH_FORMAT_CONST(OUTLINE),
    GLYPH_FORMAT_CONST(PLOTTER),
    {NULL}
};


static PyTypeObject Py_FT_RENDER_MODE_Type;
#define RENDER_MODE_CONST(name) DEF_CONST(name, FT_RENDER_MODE)
static constant_def FT_RENDER_MODE_constants[] = {
    RENDER_MODE_CONST(NORMAL),
    RENDER_MODE_CONST(LIGHT),
    RENDER_MODE_CONST(MONO),
    RENDER_MODE_CONST(LCD),
    RENDER_MODE_CONST(LCD_V),
    {NULL}
};


int setup_Glyph(PyObject *m)
{
    memset(&Py_Glyph_Type, 0, sizeof(PyTypeObject));
    Py_Glyph_Type = (PyTypeObject) {
        .tp_name = "freetypy.Glyph",
        .tp_basicsize = sizeof(Py_Glyph),
        .tp_dealloc = (destructor)Py_Glyph_dealloc,
        .tp_doc = doc_Glyph__init__,
        .tp_methods = Py_Glyph_methods,
        .tp_getset = Py_Glyph_getset,
        .tp_init = (initproc)Py_Glyph_init,
        .tp_new = Py_Glyph_new
    };

    ftpy_setup_type(m, &Py_Glyph_Type);

    if (define_constant_namespace(
            m, &Py_FT_GLYPH_BBOX_Type, &Py_FT_GLYPH_BBOX_ConstantType,
            "freetypy.GLYPH_BBOX",
            doc_GLYPH_BBOX, FT_GLYPH_BBOX_constants) ||

        define_constant_namespace(
            m, &Py_FT_GLYPH_FORMAT_Type, &Py_FT_GLYPH_FORMAT_ConstantType,
            "freetypy.GLYPH_FORMAT",
            doc_GLYPH_FORMAT, FT_GLYPH_FORMAT_constants) ||

        define_constant_namespace(
            m, &Py_FT_RENDER_MODE_Type, &Py_FT_RENDER_MODE_ConstantType,
            "freetypy.RENDER_MODE",
            doc_RENDER_MODE, FT_RENDER_MODE_constants)) {
        return -1;
    }

    return 0;
}

File: docstrings/glyph.py
# -*- coding: utf-8 -*-

# Copyright (c) 2015, Michael Droettboom All rights reserved.

# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:

# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in
#    the documentation and/or other materials provided with the
#    distribution.

# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

# The views and conclusions contained in the software and
# documentation are those of the authors and should not be interpreted
# as representing official policies, either expressed or implied, of
# the FreeBSD Project.

from __future__ import print_function, unicode_literals, absolute_import

Glyph__init__ = """
Contains information pertaining to a single glyph.

Notes
-----
If `Face.load_glyph` is called with default flags (see `LOAD.DEFAULT`)
the glyph image is loaded in the glyph slot in its native format
(e.g., an outline glyph for TrueType and Type 1 formats).

This image can later be converted into a bitmap by calling
`Glyph.render`.
"""

Glyph_advance = """
The transformed advance width for the glyph (in fractional
pixels). May not be transformed, depending on the value of
`LOAD.IGNORE_TRANSFORM`. As specified with `LOAD.VERTICAL_LAYOUT`, it
uses either the `horiAdvance` or the `vertAdvance` value of `metrics`
field.
"""

Glyph_bitmap = """
Get the `Bitmap` from this `Glyph` if it has been rendered.
"""

Glyph_bitmap_left = """
This is the bitmap's left bearing expressed in integer pixels. Of
course, this is only valid if the format is `GLYPH_FORMAT.BITMAP`.
"""

Glyph_bitmap_top = """
This is the bitmap's top bearing expressed in integer pixels. Remember
that this is the distance from the baseline to the top-most glyph
scanline, upwards y coordinates being positive. Of course, this is
only valid if the format is `GLYPH_FORMAT.BITMAP`.
"""

Glyph_face = """
The parent face object.
"""

Glyph_format = """
This field indicates the format of the image contained in the glyph
slot. Typically `GLYPH_FORMAT.BITMAP`, `GLYPH_FORMAT.OUTLINE`, or
`GLYPH_FORMAT.COMPOSITE`, but others are possible.

See `GLYPH_FORMAT` for the set of available formats.
"""

Glyph_get_cbox = """
Return a glyph's ‘control box’. The control box encloses all the
outline's points, including Bézier control points. Though it coincides
with the exact bounding box for most glyphs, it can be slightly larger
in some situations (like when rotating an outline which contains
Bézier outside arcs).

Computing the control box is very fast, while getting the bounding box
can take much more time as it needs to walk over all segments and arcs
in the outline.

Parameters
----------
mode : int, optional
    The mode which indicates how to interpret the returned bounding
    box values.  For the available options, see `GLYPH_BBOX`.

Returns
-------
bbox : BBox
    The glyph coordinate bounding box.

Notes
-----
Coordinates are relative to the glyph origin, using the *y* upwards
convention.

If the glyph has been loaded with `LOAD.NO_SCALE`, `mode` must be set
to `GLYPH_BBOX.UNSCALED` to get unscaled font units in 26.6
fixed-point pixel format.

If the font is tricky and the glyph has been loaded with
`LOAD.NO_SCALE`, the resulting `BBox` is meaningless. To get
reasonable values for the `BBox` it is necessary to load the glyph at
a large ppem value (so that the hinting instructions can properly
shift and scale the subglyphs), then extracting the `BBox` which can
be eventually converted back to font units.

Note that the maximum coordinates are exclusive, which means that one
can compute the width and height of the glyph image as::

  width  = bbox.xMax - bbox.xMin
  height = bbox.yMax - bbox.yMin

Note also that for 26.6 coordinates, if `mode` is set to
`GLYPH_BBOX.GRIDFIT`, the coordinates will also be grid-fitted, which
corresponds to::

  bbox.xMin = floor(bbox.xMin)
  bbox.yMin = floor(bbox.yMin)
  bbox.xMax = ceil(bbox.xMax)
  bbox.yMax = ceil(bbox.yMax)

To get the bbox in pixel coordinates, set `mode` to
`GLYPH_BBOX.TRUNCATE`.

To get the bbox in grid-fitted pixel coordinates, set `mode` to
`GLYPH_BBOX.PIXELS`.
"""

Glyph_linear_hori_advance = """
The advance width of the unhinted glyph. Its value is expressed in
fractional pixels, unless `LOAD.LINEAR_DESIGN` is set when loading the
glyph. This field can be important to perform correct WYSIWYG
layout. Only relevant for outline glyphs.
"""

Glyph_linear_vert_advance = """
The advance height of the unhinted glyph. Its value is expressed in
fractional pixels, unless `LOAD.LINEAR_DESIGN` is set when loading the
glyph. This field can be important to perform correct WYSIWYG
layout. Only relevant for outline glyphs.
"""

Glyph_lsb_delta = """
The difference between hinted and unhinted left side bearing while
autohinting is active. Zero otherwise.
"""

Glyph_metrics = """
The `GlyphMetrics` of the glyph. The returned values depend on the
last load flags (see the `Face.load_glyph` method) and can be
expressed either in 26.6 fractional pixels or font units.

Note that even when the glyph image is transformed, the metrics are
not.
"""

Glyph_outline = """
The outline descriptor for the current glyph image if its format is
`GLYPH_FORMAT.OUTLINE`. Once a glyph is loaded, `outline` can be
transformed, distorted, embolded, etc.
"""

Glyph_render = """
Convert a given glyph image to a `Bitmap`. It does so by inspecting
the glyph image format, finding the relevant renderer, and invoking
it.

Parameters
----------
render_mode : int, optional
    This is the render mode used to render the glyph image into a
    bitmap.

    See `RENDER_MODE` for the available options.

origin : 2-sequence of floats
    The (x, y) origin to translate the glyph image before rendering.

Returns
-------
bitmap : Bitmap
    The generated bitmap.

Notes
-----
The selected render mode only affects vector glyphs of a
font. Embedded bitmaps often have a different pixel mode like
`PIXEL_MODE.MONO`. You can use `Bitmap.convert` to transform them
into 8-bit pixmaps.
"""

Glyph_rsb_delta = """
The difference between hinted and unhinted right side bearing while
autohinting is active. Zero otherwise.
"""

Glyph_subglyphs = """
A sequence of all of the subglyphs that make up the glyph.

Empty unless `format` is `GLYPH_FORMAT.COMPOSITE`.
"""

GLYPH_BBOX = """
The mode how the values of `Glyph.get_cbox` are returned.

- `UNSCALED`: Return unscaled font units.

- `SUBPIXELS`: Return unfitted fractional coordinates.

- `GRIDFIT`: Return grid-fitted factional coordinates.

- `TRUNCATE`: Return coordinates in integer pixels.

- `PIXELS`: Return grid-fitted pixel coordinates.
"""

GLYPH_FORMAT = """
An enumeration type used to describe the format of a given glyph
image. Note that this version of FreeType only supports two image
formats, even though future font drivers will be able to register
their own format.

- `NONE`: The value 0 is reserved.

- `COMPOSITE`: The glyph image is a composite of several other
  images. This format is only used with `LOAD.NO_RECURSE`, and is
  used to report compound glyphs (like accented characters).

- `BITMAP`: The glyph image is a bitmap, and can be described as an
  `Bitmap`.

- `OUTLINE`: The glyph image is a vectorial outline made of line
  segments and Bézier arcs; it can be described as an `Outline`.

- `PLOTTER`: The glyph image is a vectorial path with no inside and
  outside contours. Some Type 1 fonts, like those in the Hershey
  family, contain glyphs in this format. These are described as
  `Outline`, but FreeType isn't currently capable of rendering them
  correctly.
"""

RENDER_MODE = """
A set of enumerations to select a bitmap renderer.

- `NORMAL`: This is the default render mode; it corresponds to 8-bit
  anti-aliased bitmaps.

- `LIGHT`: This is equivalent to `RENDER_MODE.NORMAL`. It is only
  defined as a separate value because render modes are also used
  indirectly to define hinting algorithm selectors. See
  `LOAD.TARGET_XXX` for details.

- `MONO`: This mode corresponds to 1-bit bitmaps (with 2 levels of
  opacity).

- `LCD`: This mode corresponds to horizontal RGB and BGR sub-pixel
  displays like LCD screens. It produces 8-bit bitmaps that are 3
  times the width of the original glyph outline in pixels, and which
  use the `PIXEL_MODE.LCD` mode.  On many freetype builds, this
  functionality will be disabled due to patent restrictions, in which
  case the resulting bitmap will be grayscale.

- `LCD_V`: This mode corresponds to vertical RGB and BGR sub-pixel
  displays (like PDA screens, rotated LCD displays, etc.). It produces
  8-bit bitmaps that are 3 times the height of the original glyph
  outline in pixels and use the `PIXEL_MODE.LCD_V` mode. On many
  freetype builds, this functionality will be disabled due to patent
  restrictions, in which case the resulting bitmap will be grayscale.
"""
